features/nfc/source/nfc/NFCController.cpp

```diff
@@ -0,0 +1,238 @@
+{
+    return _transceiver;
+}
+
+void NFCController::polling_callback(nfc_err_t ret)
+{
+    // Polling has completed
+    _discovery_running = false;
+
+    NFC_DBG("Polling finished with result %u", ret);
+
+    if (ret == NFC_OK) {
+        // Check if a remote initiator was detected and if so, instantiate it
+        if (!transceiver_is_initiator_mode(_transceiver)) {
+            nfc_tech_t active_tech = transceiver_get_active_techs(_transceiver);
+            if (active_tech.nfc_iso_dep_a || active_tech.nfc_iso_dep_b) {
+                Type4RemoteInitiator *type4_remote_initiator_ptr = new (std::nothrow) Type4RemoteInitiator(this, _ndef_buffer);
+                if (type4_remote_initiator_ptr != NULL) {
+                    SharedPtr<NFCRemoteInitiator> type4_remote_initiator(type4_remote_initiator_ptr);
+                    if (_delegate != NULL) {
+                        _delegate->on_nfc_initiator_discovered(type4_remote_initiator);
+                    }
+                }
+            }
+        }
+    }
+
+    if (_delegate != NULL) {
+        Delegate::nfc_discovery_terminated_reason_t reason;
+
+        // Map reason
+        switch (ret) {
+            case NFC_OK:
+                reason = Delegate::nfc_discovery_terminated_completed;
+                break;
+            case NFC_ERR_ABORTED:
+                reason = Delegate::nfc_discovery_terminated_canceled;
+                break;
+            default:
+                // Any other error code means there was an error during the discovery process
+                reason = Delegate::nfc_discovery_terminated_rf_error;
+                break;
+        }
+
+        _delegate->on_discovery_terminated(reason);
+    }
+}
+
+void NFCController::scheduler_process(bool hw_interrupt)
+{
+    _timeout.detach(); // Cancel timeout - if it triggers, it's ok as we'll have an "early" iteration which will likely be a no-op
+
+    // Process stack events
+    uint32_t timeout = nfc_scheduler_iteration(_scheduler, hw_interrupt ? EVENT_HW_INTERRUPT : EVENT_NONE);
+
+    _timeout.attach(callback(this, &NFCController::on_timeout), timeout);
+}
+
+void NFCController::on_hw_interrupt()
+{
+    // Run scheduler - this is called in interrupt context
+    _timeout.detach(); // Cancel timeout - if it triggers anyways, it's ok
+    _queue->call(this, &NFCController::scheduler_process, true);
+}
+
+void NFCController::on_timeout()
+{
+    // Run scheduler - this is called in interrupt context
+    _queue->call(this, &NFCController::scheduler_process, false);
+}
+
+void NFCController::s_polling_callback(nfc_transceiver_t *pTransceiver, nfc_err_t ret, void *pUserData)
+{
+    NFCController *self = (NFCController *) pUserData;
+    self->polling_callback(ret);
+}
+
+// Implementation nfc_scheduler_timer_t
+void nfc_scheduler_timer_init(nfc_scheduler_timer_t *timer)
+{
+    (void)timer; // This is a no-op
+}
+
+void nfc_scheduler_timer_start(nfc_scheduler_timer_t *timer)
```

Comment:
It is an mbed Timer, isn't it. The stack is not some shared library so why not updated the timer to be of type mbed::Timer?

Reply:
The stack could potentially be used elsewhere, hence for the abstract type

Reply:
To `mbed::Timer` 😕?